### PR TITLE
Improve intermediate process

### PR DESCRIPTION
**Pull request by zyiou**
based on after-merge comments on PR #51

### CODE REVIEW

File: pkg/intermediate/aggregate.go
Comment (srikartati) [inline pkg/intermediate/aggregate.go:189]: Can these lines be simplified for code (including till L221) readability purposes? 
Define the flow key elements in string list, and then iterate over them using switch-case?
There is common code for (srcPort, dstPort) and (srcIP, dstIP). This decreases the lines of code.
What do you think?

File: pkg/intermediate/aggregate_test.go
Comment (srikartati) [inline pkg/intermediate/aggregate_test.go:213]: Maybe a comment here saying that `Start` of `aggregationProcess` is blocked until the above go routine with `Stop` finishes is better to have. This will give an idea of their usage.
Proper usage would be having `Start` in a go-routine with some external channel. And until the external channel is closed, `Stop` won't be called.

File: pkg/intermediate/aggregate.go
Comment (zyiou) [reply pkg/intermediate/aggregate.go]: Sure. Updated.

File: pkg/intermediate/aggregate.go
Comment (zyiou) [reply pkg/intermediate/aggregate.go]: Agree. Changed to switch-case. Thanks!

File: pkg/intermediate/aggregate.go
Comment (shihhaoli) [inline pkg/intermediate/aggregate.go:140]: This is still inside a lib function (i.e. not an external caller).
Do we need to make a copy of the map?
Can we just lock/unlock it?

File: pkg/intermediate/aggregate.go
Comment (shihhaoli) [inline pkg/intermediate/aggregate.go:91]: We may consider returning a pointer of the new map to avoid making a second copy during returned-by-value?

File: pkg/intermediate/aggregate.go
Comment (zyiou) [reply pkg/intermediate/aggregate.go]: Sure. Updated. Thanks!

File: pkg/intermediate/aggregate.go
Comment (zyiou) [reply pkg/intermediate/aggregate.go:91]: [Map](https://blog.golang.org/maps) in Golang is reference type like pointer, so it should be returned by reference.

Comment (shihhaoli) [reply pkg/intermediate/aggregate.go:91]: Cool, nice to learn that. Thanks.

Comment (srikartati) [reply pkg/intermediate/aggregate.go:91]: The main objective of this function is to iterate over the records and perform a common operation.
Could we do something like below in Antrea to avoid the step of creating a copy of the map? I think this map will be common to the cluster w.r.t flow aggregator implementation and could be very large; making a copy can involve high overhead.

https://github.com/vmware-tanzu/antrea/blob/master/pkg/agent/flowexporter/connections/connections.go#L209

Comment (srikartati) [reply pkg/intermediate/aggregate.go:91]: I see the above function is currently utilized only in tests, and all the others are internal functions that access the map directly. However, exposing this function may lead to inefficient usage.

Comment (zyiou) [reply pkg/intermediate/aggregate.go:91]: Thanks for the suggestion. For flow aggregator, we will need to get the map, obtain the records and send them, that is the use case for exposing the function.

File: pkg/intermediate/aggregate.go
Comment (zyiou) [reply pkg/intermediate/aggregate.go:204]: I was thinking separate sourceIPv6Address and destinationIPv6Address would cause duplications. But it is clearer. Thanks!

File: pkg/intermediate/aggregate.go
Comment (zyiou) [reply pkg/intermediate/aggregate.go:91]: Checked with @dreamtalen that passing a callback function will be enough for flow aggregator to process the records.

**Review by srikartati — COMMENTED** [review]
LGTM. 
Please add some comments for methods in the Aggregation data structure. Especially where lock needs to be acquired by the caller. 
Also, changing name from aggregation->aggregationProcess is better? We can be consistent with exportingProcess and collectionProcess.

**Comment by zyiou** [comment]
> Please add some comments for methods in the Aggregation data structure. Especially where lock needs to be acquired by the caller.

Currently the lock will not be acquired publicly. Which part are you referring to?
> Also, changing name from aggregation->aggregationProcess is better? We can be consistent with exportingProcess and collectionProcess.

Done. Thanks!

**Comment by srikartati** [comment]
> > Please add some comments for methods in the Aggregation data structure. Especially where lock needs to be acquired by the caller.
> 
> Currently the lock will not be acquired publicly. Which part are you referring to?

I was referring to `addRecordToMap`. Yes, it is an internal method, but the add function of the map is usually protected with a lock. If someone wants to extend the aggregateProcess structure in future, comments will be useful.

**Comment by zyiou** [comment]
> I was referring to `addRecordToMap`. Yes, it is an internal method, but the add function of the map is usually protected with a lock. If someone wants to extend the aggregateProcess structure in future, comments will be useful.

Got it. Thanks!

**Review by srikartati — DISMISSED** [review]
Thanks for working on this.
LGTM other than comments for methods. 
Please merge once you get approval from Shihao.

File: pkg/intermediate/aggregate.go
Comment (shihhaoli) [inline pkg/intermediate/aggregate.go:238]: Do we want to print warning if both (src ipv4 and src ipv6) or (dst ipv4 and dst ipv6) are set?

Comment (zyiou) [reply pkg/intermediate/aggregate.go]: Sure. Updated

**Review by shihhaoli — APPROVED** [review]
Thanks for the changes. LGTM.

**Pull request merged**
